sdk/appconfiguration/azure-spring-cloud-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/implementation/AppConfigurationPropertySourceLocator.java

```diff
@@ -0,0 +1,301 @@
+                        break;
+                    }
+                    if (generatedPropertySources) {
+                        break;
+                    }
+                }
+
+                if (generatedPropertySources) {
+                    // Updating list of propertySources
+                    sourceList.forEach(composite::addPropertySource);
+                } else if (!STARTUP.get() || (configStore.isFailFast() && STARTUP.get())) {
+                    String message = "Failed to generate property sources for " + configStore.getEndpoint();
+
+                    // Refresh failed for a config store ending attempt
+                    failedToGeneratePropertySource(configStore, newState, new RuntimeException(message));
+                }
+
+            } else if (!configStore.isEnabled() && loadNewPropertySources) {
+                LOGGER.info("Not loading configurations from {} as it is not enabled.", configStore.getEndpoint());
+            } else {
+                LOGGER.warn("Not loading configurations from {} as it failed on startup.", configStore.getEndpoint());
+            }
+        }
+
+        StateHolder.updateState(newState);
+        STARTUP.set(false);
+
+        return composite;
+    }
+
+    private List<ConfigurationSetting> getWatchKeys(AppConfigurationReplicaClient client,
+        List<AppConfigurationStoreTrigger> triggers) {
+        List<ConfigurationSetting> watchKeysSettings = new ArrayList<>();
+        for (AppConfigurationStoreTrigger trigger : triggers) {
+            ConfigurationSetting watchKey = client.getWatchKey(trigger.getKey(), trigger.getLabel());
+            if (watchKey != null) {
+                watchKeysSettings.add(watchKey);
+            } else {
+                watchKeysSettings
+                    .add(new ConfigurationSetting().setKey(trigger.getKey())
+                        .setLabel(trigger.getLabel()));
+            }
+        }
+        return watchKeysSettings;
+    }
+
+    private void getFeatureFlagWatchKeys(AppConfigurationReplicaClient client, ConfigStore configStore,
+        StateHolder newState) {
+        List<ConfigurationSetting> watchKeysFeatures = new ArrayList<>();
+
+        if (configStore.getFeatureFlags().getEnabled()) {
+            SettingSelector settingSelector = new SettingSelector()
+                .setKeyFilter(configStore.getFeatureFlags().getKeyFilter())
+                .setLabelFilter(configStore.getFeatureFlags().getLabelFilter());
+
+            PagedIterable<ConfigurationSetting> watchKeys = client.listSettings(settingSelector);
+
+            watchKeys
+                .forEach(watchKey -> watchKeysFeatures.add(NormalizeNull.normalizeNullLabel(watchKey)));
+
+            newState.setStateFeatureFlag(configStore.getEndpoint(), watchKeysFeatures,
+                configStore.getMonitoring().getFeatureFlagRefreshInterval());
+            newState.setLoadStateFeatureFlag(configStore.getEndpoint(), true);
```

Comment:
This method is called `getFeatureFlagWatchKeys`, it suggests that the getting operation has no side effects, so we probably should move the state setting action out from this method.